fl4health/clients/nnunet_client.py

```diff
@@ -0,0 +1,795 @@
+        # Issue: https://github.com/MIC-DKFZ/batchgenerators/issues/123
+        # PR: https://github.com/MIC-DKFZ/batchgenerators/pull/124
+        FLOWER_LOGGER.propagate = False
+
+        # Redirect nnunet output to flwr logger at DEBUG level
+        with redirect_stdout(self.stream2debug):
+            # Get the nnunet dataloader iterators. (Technically augmenter classes)
+            train_loader, val_loader = self.nnunet_trainer.get_dataloaders()
+
+        # Now clear the root handler that was create and turn propagate back to true
+        root_logger = logging.getLogger()
+        root_logger.handlers.clear()
+        FLOWER_LOGGER.propagate = True
+
+        # Wrap nnunet dataloaders to make them compatible with fl4health
+        train_loader = nnUNetDataLoaderWrapper(nnunet_augmenter=train_loader, nnunet_config=self.nnunet_config)
+        val_loader = nnUNetDataLoaderWrapper(nnunet_augmenter=val_loader, nnunet_config=self.nnunet_config)
+
+        if self.verbose:
+            log(INFO, f"\tDataloaders initialized in {time.time()-t:.1f}s")
+
+        return train_loader, val_loader
+
+    def get_model(self, config: Config) -> nn.Module:
+        return self.nnunet_trainer.network
+
+    def get_criterion(self, config: Config) -> _Loss:
+        return Module2LossWrapper(self.nnunet_trainer.loss)
+
+    def get_optimizer(self, config: Config) -> Optimizer:
+        return self.nnunet_trainer.optimizer
+
+    def get_lr_scheduler(self, config: Config) -> _LRScheduler:
+        """
+        Creates an LR Scheduler similar to the nnunet default except we set max steps
+        to the total number of steps and update every step. Initial and final LR are
+        the same as nnunet, difference is nnunet sets max steps to num 'epochs', but
+        they define an 'epoch' as exactly 250 steps. Therefore they update every 250
+        steps. Override this method to set your own LR scheduler.
+
+        Args:
+            config (Config): The server config. This method will look for the
+        Returns:
+            _LRScheduler: The default nnunet LR Scheduler for nnunetv2 2.5.1
+        """
+        if not isinstance(self.nnunet_trainer.lr_scheduler, PolyLRScheduler):
+            log(
+                WARNING,
+                (
+                    "Nnunet seems to have changed their default LR scheduler to "
+                    f"type: {type(self.nnunet_trainer.lr_scheduler)}. "
+                    "Using PolyLRScheduler instead. Override or update the "
+                    "get_lr_scheduler method of nnUNetClient to change this"
+                ),
+            )
+
+        # Determine total number of steps throughout all FL rounds
+        local_epochs, local_steps, _, _ = self.process_config(config)
+        if local_steps is not None:
+            self.steps_per_round = local_steps
+            self.total_steps = int(config["n_server_rounds"]) * self.steps_per_round
```

Comment:
BasicClient already has an attribute total_steps that represents the total number of steps completed up until a given point so we do not want to override that

Reply:
good catch